### PR TITLE
Add Support for Python 3.6

### PR DESCRIPTION
**Pull request by xoviat**
Okay, let's try this again. This should in theory fully support Python 3.6, although I don't think I committed properly. We'll see.

### CODE REVIEW

**Comment by xoviat** [comment]
That's better.

**Comment by xoviat** [comment]
Fixes #2331.

**Comment by xoviat** [comment]
I should note that the test stop after three failures, which is kind of bad because we are already getting failures from sphinx (PR #2323) and threading. So this code isn't fully tested here because of those unrelated failures. My suggestion would be to merge #2323  and then I can force travis to fire again.

**Comment by xoviat** [comment]
Hrm...

https://travis-ci.org/pyinstaller/pyinstaller/jobs/187940282#L1292
```
Traceback (most recent call last):
  File "scripts/pyi_python_makefile.py", line 20, in <module>
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 481, in get_config_h_filename
    inc_dir = get_path('platinclude')
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 512, in get_path
    return get_paths(scheme, vars, expand)[name]
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 502, in get_paths
    return _expand_vars(scheme, vars)
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 164, in _expand_vars
    _extend_dict(vars, get_config_vars())
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 550, in get_config_vars
    _init_posix(_CONFIG_VARS)
  File "/opt/python/3.6-dev/lib/python3.6/sysconfig.py", line 421, in _init_posix
    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
ModuleNotFoundError: No module named '_sysconfigdata_m_linux_x86_64-linux-gnu'
```
Possbly related to [this][1]?

Also, we're not importing idlelib:

https://travis-ci.org/pyinstaller/pyinstaller/jobs/187940282#L2214
```
Traceback (most recent call last):
  File "/tmp/pytest-of-travis/pytest-0/popen-gw2/test_idlelib_onedir_0/test_idlelib.py", line 4, in <module>
    import idlelib.TreeWidget
```

[1]: https://bugs.python.org/msg275650

**Comment by xoviat** [comment]
Looks like we're down to idlegraph (waiting on #2342) and sphinx ( #2323).

**Comment by htgoebel** [comment]
@xoviat I *really* appreciate your work on this. But **please** get yourself an account on travis and test you changes in your own fork (this is easy, taking about 15 minutes and gives you results without need to update the pull-request). If you are pushing your changes here every time, I get lots of messages for stuff which is not yet finished. I need to look at theme until I find out that it is still work in progress.

**Comment by xoviat** [comment]
Current status: Everyone uses pypiwin32 for the pywin32 dependency, and:
1. pypiwin32 doesn't have a binary. 
2. It is no longer possible to build pywin32 from source using the standard build tools (pywin32/pypiwin32#2). 

Fortunately, after some investigation, I have been able to `wheel convert` the released windows installer from sourceforge and have uploaded it to [the releases](https://github.com/pywin32/pypiwin32/releases). I am hoping that @glyph will have it uploaded to pypi soon.

**Comment by glyph** [comment]
@xoviat I'm not entirely sure what the next action for me is here. https://github.com/pywin32/pypiwin32/releases says "there aren't any releases here" so I can't upload those...

**Comment by xoviat** [comment]
@glyph The releases are up along with the generator I used.

**Comment by xoviat** [comment]
@htgoebel That's all the time that I have. It looks like there are some issues with a few specific libraries that still need to be solved, but those can be done with some changes to the hooks. 

What needs to be done:
- [ ] Sane installation of lxml (the following works for py36 but not all other py versions)
    - `nuget install libxml2`
    - `pip install lxml`
- [ ] pypiwin32
- [ ] Sane installation of matplotlib on windows
- [ ] idlelib on linux
- [ ] #2323 

It looks like the threading issues on linux are gone? If not, then I guess include that as well. Anyway, I hope that I made a reasonable contribution and good luck to the PyInstaller team!

**Comment by xoviat** [comment]
@htgoebel Go ahead and merge this. I have opened individual issues for the specific libraries that need to be resolved.

**Comment by smurphy917** [comment]
per @xoviat, this fixes #2393

File: appveyor.yml [hunk 1]
```diff
@@ -40,23 +40,6 @@ environment:
...
-      # for more discussion.
-      TEST1_CMD: "py.test --maxfail 5 --durations=10 tests\\unit tests\\functional -k \"not tests/functional/test_libraries.py\""
-      TEST2_CMD: ""
-
-    - PYTHON: "C:\\Python35-x64"
-      PYTHON_VERSION: "3.5"
-      PYTHON_ARCH: "64"
-      VCVARSALL: "x64"
-      TEST1_CMD: ""
-      TEST2_CMD: "py.test --maxfail 5 --durations=10 tests\\functional\\test_libraries.py"
```
Comment (htgoebel) [inline appveyor.yml]: Why did you remove this block?

File: tests/requirements-win.txt [hunk 2]
```diff
@@ -1,3 +1,5 @@
+--extra-index-url https://pypi.anaconda.org/giumas/simple
+--extra-index-url https://pypi.anaconda.org/xoviat/simple
```
Comment (htgoebel) [inline tests/requirements-win.txt]: Why do we need this here? All pacakges in this fail ought to be available at PyPI, i Thought?

Comment (xoviat) [inline tests/requirements-win.txt]: Lxml was previously installed directly in appveyor.yml, per #1638. I had to create a repository because the lxml binaries for 3.6 were not in the original repository. During the process, I thought it would be cleaner to have the URLs in a requirements file so that it's not specific to windows.

File: appveyor.yml (hunk 1 above)
Comment (xoviat) [inline appveyor.yml]: Because Python 3.6 x64 is now tested, it doesn't make sense to test 3.5 x64. The appveyor tests are long enough as it is already.

Comment (htgoebel) [inline appveyor.yml]: I take this as: We don't need to test the whole test-quite for both 64 and 32 bit. So for 64 bit we only run basic tests do make sure the bootloader is working correctly.

Comment (xoviat) [inline appveyor.yml]: I guess this was removed from the diff? Not sure what happened there, as I made no changes on my end.

**Comment by htgoebel** [comment]
Merged manually, see 1509f0b5288843cc352da19469b79c35999c1835
